internal/interface/crawler/crawler.go

```diff
@@ -92,71 +237,94 @@ func (c *Crawler) upsertArticle(article model.Article) (bool, error) {
 
 func (c *Crawler) CrawlArticles(cfg Configuration) error {
 	fmt.Println("Crawling...")
-	collector := colly.NewCollector()
-	err := collector.Limit(&colly.LimitRule{
-		DomainGlob:  "*",
-		Delay:       time.Second / 4,
-		RandomDelay: time.Second / 4,
-	})
-	if err != nil {
-		return err
-	}
-
-	urlQueue := []string{cfg.RootURL}
-	var parseErr error
 
-	collector.OnResponse(c.responseProcessor(&parseErr, &urlQueue, &cfg))
-	collector.OnRequest(func(r *colly.Request) {
-		fmt.Println("Visiting", r.URL)
-	})
+	ctx, cancel := context.WithCancel(context.Background())
+	osChan := make(chan os.Signal, 1)
+	signal.Notify(osChan, os.Interrupt)
+	defer func() {
+		signal.Stop(osChan)
+		cancel()
+	}()
 
-	articlesCnt, err := c.getArticlesCount()
-	for err == nil && articlesCnt < cfg.DesiredArticleCount && len(urlQueue) > 0 && parseErr == nil {
-		u := urlQueue[0]
-		totalURLsVisited.Inc()
-		urlQueue = urlQueue[1:]
-		timeStart := time.Now()
-		collector.Visit(u)
-		urlVisitDuration.Observe(time.Now().Sub(timeStart).Seconds())
-		articlesCnt, err = c.getArticlesCount()
-	}
-	if err != nil {
-		return err
-	}
-	return parseErr
-}
+	URLChan := make(chan string, chBuff)
+	HTMLChan := make(chan *http.Response, chBuff)
+	ArticleChan := make(chan model.Article, chBuff)
+	NewURLChan := make(chan string, chBuff)
 
-func (c *Crawler) responseProcessor(parseErr *error, urlQueue *[]string, cfg *Configuration) func(response *colly.Response) {
-	return func(response *colly.Response) {
-		dom, err := goquery.NewDocumentFromReader(bytes.NewReader(response.Body))
-		if err != nil {
-			parseErr = &err
-			return
-		}
-		err = c.collectUrls(dom, urlQueue, cfg)
-		if err != nil {
-			parseErr = &err
-			return
-		}
-		if strings.Contains(response.Request.URL.String(), "abs/") {
-			article, err := c.parseArticle(response, dom)
-			if err != nil {
-				parseErr = &err
-				return
-			}
-			up, err := c.upsertArticle(article)
-			if err != nil {
-				parseErr = &err
+	go func() {
+		for {
+			select {
+			case <-osChan:
+				cancel()
+			case <-ctx.Done():
 				return
 			}
-			if up {
-				fmt.Println("Upserted article", article.Id)
-			}
 		}
+	}()
+
+	var gwg sync.WaitGroup
+	gwg.Add(1)
+	go func(out chan<- string) {
+		err := c.GetURLFromDB(ctx, out)
+		fmt.Fprintf(os.Stderr, "URLGetter 1 stopped, reason: %s\n", err)
+		gwg.Done()
+		cancel()
+	}(URLChan)
+
+	var dwg sync.WaitGroup
+	dwg.Add(downloadURLConcurrency)
+	for i := 0; i < downloadURLConcurrency; i++ {
+		go func(in <-chan string, out chan<- *http.Response) {
+			err := c.DownloadURL(ctx, in, out)
+			fmt.Fprintf(os.Stderr, "URLDownloader %d stopped, reason: %s\n", i, err)
+			dwg.Done()
+			cancel()
```

Comment:
What do you mean by calling cancel function here?

Reply:
We cancel the whole pipeline if any error occurs at any point of processing. Of course we should treat errors like failed http download differently and only cancel everything if we have encountered a situation that we cannot respond to properly, but at this point it seems to be good enough.